components/org.wso2.identity.event.websubhub.publisher/src/main/java/org/wso2/identity/event/websubhub/publisher/service/WebSubEventSubscriberImpl.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.http.HttpStatus;
+import org.apache.http.StatusLine;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.wso2.carbon.identity.webhook.management.api.exception.WebhookMgtException;
+import org.wso2.carbon.identity.webhook.management.api.service.WebhookSubscriber;
+import org.wso2.identity.event.websubhub.publisher.constant.WebSubHubAdapterConstants;
+import org.wso2.identity.event.websubhub.publisher.exception.WebSubAdapterException;
+import org.wso2.identity.event.websubhub.publisher.internal.ClientManager;
+import org.wso2.identity.event.websubhub.publisher.internal.WebSubHubAdapterDataHolder;
+import org.wso2.identity.event.websubhub.publisher.util.WebSubHubCorrelationLogUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.buildSubscriptionURL;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.constructHubTopic;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.getWebSubBaseURL;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.handleErrorResponse;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.handleFailedOperation;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.handleSuccessfulOperation;
+
+/**
+ * OSGi service for managing WebSub Hub subscriptions.
+ */
+public class WebSubEventSubscriberImpl implements WebhookSubscriber {
+
+    private static final Log log = LogFactory.getLog(WebSubEventSubscriberImpl.class);
+
+    @Override
+    public String getName() {
+
+        return WebSubHubAdapterConstants.WEB_SUB_HUB_ADAPTER_NAME;
+    }
+
+    @Override
+    public boolean subscribe(List<String> topics, String callbackUrl, String tenantDomain) {
+
+        try {
+            // Create a list to hold all the futures
+            List<CompletableFuture<Boolean>> futures = new ArrayList<>();
+
+            for (String topic : topics) {
+                futures.add(CompletableFuture.supplyAsync(() -> {
+                    try {
+                        makeSubscriptionAPICall(constructHubTopic(topic, tenantDomain),
+                                getWebSubBaseURL(),
+                                WebSubHubAdapterConstants.Http.SUBSCRIBE,
+                                callbackUrl);
+                        log.debug("WebSub Hub subscription successful for topic: " + topic +
+                                " with callback URL: " + callbackUrl + " in tenant: " + tenantDomain);
+                        return true;
+                    } catch (WebSubAdapterException e) {
+                        log.error("Error subscribing to topic: " + topic, e);
+                        return false;
```

Comment:
Ideally you need to populate a list of topics that have failed and it should be logged proerly.

More over how do make sure if any of these are failed the overall call fails ?
Will line 93 check that